warehouse/oidc/models/gitlab.py

```diff
@@ -0,0 +1,288 @@
+        "environment": _check_environment,
+    }
+
+    __unchecked_claims__ = {
+        "namespace_id",
+        "namespace_path",
+        "user_id",
+        "user_login",
+        "user_email",
+        "user_identities",
+        "pipeline_id",
+        "pipeline_source",
+        "job_id",
+        "ref",
+        "ref_type",
+        "ref_protected",
+        "environment_protected",
+        "deployment_tier",
+        "environment_action",
+        "runner_id",
+        "runner_environment",
+        "ci_config_sha",
+        "project_visibility",
+        "jti",
+    }
+
+    @staticmethod
+    def __lookup_all__(klass, signed_claims: SignedClaims) -> Query | None:
+        # This lookup requires the environment claim to be present;
+        # if it isn't, bail out early.
+        if not (environment := signed_claims.get("environment")):
+            return None
+
+        project_path = signed_claims["project_path"]
+        ci_config_ref_prefix = f"gitlab.com/{project_path}//"
+        ci_config_ref = signed_claims["ci_config_ref_uri"].removeprefix(
+            ci_config_ref_prefix
+        )
+        namespace, project = project_path.rsplit("/", 1)
+
+        return (
+            Query(klass)
+            .filter_by(
+                namespace=namespace,
+                project=project,
+                environment=environment,
+            )
+            .filter(
+                literal(ci_config_ref).like(func.concat(klass.workflow_filepath, "%"))
+            )
+        )
+
+    @staticmethod
+    def __lookup_no_environment__(klass, signed_claims: SignedClaims) -> Query | None:
+        project_path = signed_claims["project_path"]
+        ci_config_ref_prefix = f"gitlab.com/{project_path}//"
+        ci_config_ref = signed_claims["ci_config_ref_uri"].removeprefix(
+            ci_config_ref_prefix
+        )
+        namespace, project = project_path.rsplit("/", 1)
+
+        return (
+            Query(klass)
+            .filter_by(
+                namespace=namespace,
+                project=project,
+                environment="",
+            )
+            .filter(
+                literal(ci_config_ref).like(func.concat(klass.workflow_filepath, "%"))
+            )
+        )
+
+    __lookup_strategies__ = [
+        __lookup_all__,
+        __lookup_no_environment__,
+    ]
+
+    @property
+    def project_path(self):
+        return f"{self.namespace}/{self.project}"
+
+    @property
+    def sub(self):
+        return f"project_path:{self.project_path}"
+
+    @property
+    def ci_config_ref_uri(self):
+        return f"gitlab.com/{self.project_path}//{self.workflow_filepath}"
+
+    @property
+    def publisher_name(self):
+        return "GitLab"
+
+    def publisher_url(self, claims=None):
+        base = f"https://gitlab.com/{self.project_path}"
+        sha = claims.get("sha") if claims else None
+
+        if sha:
+            return f"{base}/commit/{sha}"
+        return base
```

Comment:
Should we do something else if `sha` isn't present but `ref_path` is?

Reply:
This code actually only checks if `claims` is present. We assume that both `sha` and `ref_path` should be present (see my comment below)

Reply:
I think this is the only thing still outstanding:

```suggestion
    def publisher_url(self, claims=None):
        base = f"https://gitlab.com/{self.project_path}"
        return f"{base}/commit/{claims['sha']}" if claims else base
```